alacritty_terminal/src/renderer/rects.rs

```diff
@@ -103,7 +103,11 @@ impl Rects {
 
             // Check if there's an active line
             if let Some(line) = self.inner.get_mut(flag).and_then(|lines| lines.last_mut()) {
-                if cell.line == line.start.line && cell.fg == line.color {
+                let line_end = line.start.col.0 + (line.rect.width / size.cell_width) as usize;
```

Comment:
Would it be better to store this `end` on the `Line`? We already store the `start` in addition to the underlying `Rect`, so maybe this would be a natural addition.

I will say however, I'm not overly in love with storing a `Rect` with it's own positions independently of a `start` and `end` point, but I can see why it might be needed for performance reasons. It might be a good idea to add a note on the `start` (and potential `end`) fields explaining the constraint with the `rect.x` & `rect.width` values.

Reply:
> I will say however, I'm not overly in love with storing a Rect with it's own positions independently of a start and end point, but I can see why it might be needed for performance reasons.

My thought was similar, but if we store both start and end, we can just transform it to rects in the `rects()` method. There's no performance benefit to creating the rects once versus creating them every time `rects()` is called, since the method is only called once per draw call.

Generally I'm not a super huge fan of the current code, since it seems vulnerable to slight mathematical errors. A cast to `usize` could easily reduce the line length by a first cell just because it's been off by like `.00000001`.

I haven't had a ton of time yet today to explore alternatives, but storing the `end` and transforming in the `rects()` call was the best thing I could come up with at this point.

Reply:
We're storing rectangles other than underline/strikeout in our `HashMap<Flags, Vec<Lines>>`, so creating such rectangles will become less  flexible (e.g. `push` method on `Rects`).

However I have enough time today to implement all of these stuff, so I'll send 2 commits with just adding `end: Point` and with refactoring you've suggested here.

Reply:
Yea, the `push` method is kinda a sore point for this code at the moment. For the purposes of this fix we don't need to go crazy refactoring things, but at some point we may want to split up the line functions from the more general rect functions in a way that addresses our concerns @chrisduerr.

Reply:
> Yea, the push method is kinda a sore point for this code at the moment. But for the purposes of this fix we don't need to go crazy refactoring things. At some point we may want to split up the line functions from the more general rect functions in a way that addresses our concerns @chrisduerr.

Well we've already refactored things a lot, so I'm not too concerned personally.

However I do question if there would be much benefit to separating them more.

However, we never need to call `update_lines` after underline and strikeout are finished, so I'd be perfectly fine with moving `update_lines` on `Line` maybe, then having the `rects` method take `rects(self)` and return an owned rects array where we can just push new rects in.

I think this should be a minor change which could also separate concerns a bit better? Curious about your opinions on this @kchibisov / @nixpulvis.

Reply:
> I think this should be a minor change which could also separate concerns a bit better? Curious about your opinions on this @kchibisov / @nixpulvis.

I'll try to split things more, I just don't want to touch render with these refactorings.

Reply:
> I'll try to split things more, I just don't want to touch renderer with these refactorings.

Completely agree with that. At the end it should spit out a `Vec<Rects>` and we should be good. Though some minor changes in `display.rs` should be fine.

I think making the underline/strikeout solely handled by the `Line` struct which is then converted to a `Vec<Rect>` is the way to go. With Rust's ownership semantics that should provide a very nice interface when `rects` moves `self`.